external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,445 @@
+    }
+
+    // ====== Private helper methods ======
+
+    private void commitOffsetsForAckedTuples() {
+        final Map<TopicPartition, OffsetAndMetadata> nextCommitOffsets = new HashMap<>();
+
+        try {
+            for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                final OffsetAndMetadata offsetAndMetadata = tpOffset.getValue().findNextCommitOffset();
+                if (offsetAndMetadata != null) {
+                    nextCommitOffsets.put(tpOffset.getKey(), offsetAndMetadata);
+                }
+            }
+
+            if (!nextCommitOffsets.isEmpty()) {
+                kafkaConsumer.commitSync(nextCommitOffsets);
+                LOG.debug("Offsets successfully committed to Kafka [{}]", nextCommitOffsets);
+                // Instead of iterating again, we could commit and update state for each TopicPartition in the prior loop,
+                // but the multiple network calls should be more expensive than iterating twice over a small loop
+                for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                    final OffsetEntry offsetEntry = tpOffset.getValue();
+                    offsetEntry.updateAckedState(nextCommitOffsets.get(tpOffset.getKey()));
+                }
+            } else {
+                LOG.trace("No offsets to commit. {}", toString());
+            }
+        } catch (Exception e) {
+            LOG.error("Exception occurred while committing to Kafka offsets of acked tuples", e);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "{acked=" + acked + "} ";
+    }
+
+    // ======= Offsets Commit Management ==========
+
+    private static class OffsetComparator implements Comparator<MessageId> {
+        public int compare(MessageId m1, MessageId m2) {
+            return m1.offset() < m2.offset() ? -1 : m1.offset() == m2.offset() ? 0 : 1;
+        }
+    }
+
+    /** This class is not thread safe */
+    private class OffsetEntry {
+        private final TopicPartition tp;
+        private long committedOffset;               // last offset committed to Kafka, or initial fetching offset (initial value depends on offset strategy. See KafkaSpoutConsumerRebalanceListener)
+        private final NavigableSet<MessageId> ackedMsgs = new TreeSet<>(OFFSET_COMPARATOR);     // acked messages sorted by ascending order of offset
+
+        public OffsetEntry(TopicPartition tp, long committedOffset) {
+            this.tp = tp;
+            this.committedOffset = committedOffset;
+            LOG.debug("Created OffsetEntry for [topic-partition={}, committed-or-initial-fetch-offset={}]", tp, committedOffset);
+        }
+
+        public void add(MessageId msgId) {          // O(Log N)
+            ackedMsgs.add(msgId);
+        }
+
+        /**
+         * This method has side effects. The method updateAckedState should be called after this method.
```

Comment:
Actually where are the side effects?  I don't see any member variable being modified by the method.